vllm/v1/spec_decode/eagle_proposer.py

```diff
@@ -0,0 +1,273 @@
+# SPDX-License-Identifier: Apache-2.0
+import itertools
+
+import torch
+from torch import Tensor, nn
+
+from vllm.config import VllmConfig
+from vllm.forward_context import set_forward_context
+from vllm.v1.attention.backends.flash_attn import FlashAttentionMetadata
+from vllm.v1.outputs import SamplerOutput
+from vllm.v1.sample.metadata import SamplingMetadata
+
+
+class EagleProposer:
+
+    def __init__(self, vllm_config: VllmConfig, model: nn.Module,
+                 sampling_metadata: SamplingMetadata):
+        """
+        Initialize EagleProposer with necessary configuration and model.
+        """
+        self._vllm_config = vllm_config
+        self._model = model
+        self._sampling_metadata = sampling_metadata
+
+    def generate_draft_token_ids(
+            self, *, target_model_input_ids: Tensor,
+            target_model_positions: Tensor, target_model_hidden_states: Tensor,
+            target_model_seq_lens: list[int],
+            sampled_token_ids: list[list[int]],
+            next_prompt_token_ids: list[list[int]], is_prefill: list[bool],
+            num_draft_tokens_to_propose: int,
+            attention_metadata: FlashAttentionMetadata) -> list[SamplerOutput]:
+        """
+        Generates speculative draft token IDs using the Eagle model.
+
+        This function aligns the Eagle model's KV cache with the target
+        model’s output before generating speculative tokens. It first
+        performs a prefill pass, followed by iterative speculation passes
+        to generate additional draft tokens.
+
+        Example:
+        Consider three sequences:
+        - S1: A completed prefill sequence starting at position 0.
+        - S2: A partial prefill sequence starting at a nonzero position.
+        - S3: A decoding sequence.
+
+        Target Model:
+        Sequences: [S1, S2, S3]
+        Tokens: [T11, T12, T13, T14, T21, T22, T23, T31, T32, T33]
+        Positions: [0, 1, 2, 3, 9, 10, 11, 44, 45, 46]
+        Hidden States: [H11, H12, H13, H14, H21, H22, H23, H31, H32, H33]
+        Sampled Tokens: [[T15], [], [T32]]
+        Next Prompt Tokens: [[], [T24], []]
+
+        The first forward pass in Eagle aligns its KV cache with that
+        of the target model and generated the first proposal. The
+        input to the eagle model for the first forward pass will be
+        the following.
+
+        Eagle Prefill Forward Pass:
+        Sequences: [S1, S2, S3]
+        Tokens: [T12, T13, T14, T15, T22, T23, T24, T32]
+        Positions: [0, 1, 2, 3, 9, 10, 11, 44]
+        Previous Hidden States: [H11, H12, H13, H14, H21, H22, H23, H31]
+        Sampled Tokens: [[T16], [T25], [T33']]
```

Comment:
Minor: why is this example different from the example of input? Maybe just use the same example?

Reply:
It is the same example IIUC. The above one shows all the information including inputs and outputs, while this part only shows the inputs.